### PR TITLE
CB-17571 - Implement core RDS upgrade flow skeleton required on Azure

### PR DESCRIPTION
**Pull request by pdarvasi**
This commit:
 - implements the wire of the core RDS upgrade flow
 - implements the business logic of the `backupRds` and `restoreRds` actual flow steps
 - implements stackV4 and distroX upgradeRds endpoints

What is yet to come
 - business logic of the `stop` and `start` and `upgradeRds` actual flow steps
 - Return RDS engine version via Redbeams API

The related tasks:
[CB-17761](https://jira.cloudera.com/browse/CB-17761)	Implement database upgrade by replacing the RDS	
[CB-17762](https://jira.cloudera.com/browse/CB-17762)	Return RDS engine version via Redbeams API	
[CB-17763](https://jira.cloudera.com/browse/CB-17763)	Implement stop and start flow steps	

FD499299-ECA0-477A-A0E8-DC33D1862AAB

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/rds/upgrade/UpgradeRdsService.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import com.sequenceiq.cloudbreak.api.endpoint.v4.common.DetailedStackStatus;
+import com.sequenceiq.cloudbreak.cloud.store.InMemoryStateStore;
+import com.sequenceiq.cloudbreak.core.flow2.stack.CloudbreakFlowMessageService;
+import com.sequenceiq.cloudbreak.event.ResourceEvent;
+import com.sequenceiq.cloudbreak.orchestrator.exception.CloudbreakOrchestratorException;
+import com.sequenceiq.cloudbreak.service.StackUpdater;
+import com.sequenceiq.cloudbreak.service.upgrade.rds.RdsUpgradeOrchestratorService;
+
+@Service
+public class UpgradeRdsService {
```
Comment (horadla23) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/rds/upgrade/UpgradeRdsService.java:24]: UT would be nice for this

**Review by horadla23 — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/rds/upgrade/UpgradeRdsService.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        setStatusAndNotify(stackId, "Upgrading database server", ResourceEvent.CLUSTER_RDS_UPGRADE_DBSERVER_UPGRADE);
+    }
+
+    void restoreRdsState(Long stackId) {
+        setStatusAndNotify(stackId, "Restoring data from the backup, it might take a while", ResourceEvent.CLUSTER_RDS_UPGRADE_RESTORE_DATA);
+    }
+
+    void startServicesState(Long stackId) {
+        setStatusAndNotify(stackId, "Starting back Cloudera Manager and Runtime Services", ResourceEvent.CLUSTER_RDS_UPGRADE_START_SERVICES);
+    }
```
[the remark this answers is not in the document]
Comment (pdarvasi) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/rds/upgrade/UpgradeRdsService.java:55]: These are not intended for public consumption.

[the remark this answers is not in the document]
Comment (horadla23) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/rds/upgrade/UpgradeRdsService.java]: 👍

**Review by horadla23 — APPROVED** [review]

File: core/src/main/java/com/sequenceiq/cloudbreak/service/database/DatabaseService.java [hunk 3]
```diff
@@ -36,13 +37,13 @@ public class DatabaseService {
     @Inject
     private DatabaseServerConverter databaseServerConverter;
 
-    public StackDatabaseServerResponse getDatabaseServer(String clusterCrn) {
-        Stack stack = stackOperations.getStackByCrn(clusterCrn);
+    public StackDatabaseServerResponse getDatabaseServer(NameOrCrn nameOrCrn) {
```
Comment (doktoric) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/database/DatabaseService.java:40]: unit test would be great here

**Review by bergerdenes — CHANGES_REQUESTED** [review]
Minor stuff

File: core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/StackV4Endpoint.java [hunk 4]
```diff
@@ -432,6 +435,15 @@ FlowIdentifier upgradeClusterByNameInternal(@PathParam("workspaceId") Long works
...
             @QueryParam("imageId") String imageId, @QueryParam("initiatorUserCrn")  String initiatorUserCrn);
 
+    @PUT
+    @Path("internal/{name}/rds_upgrade")
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = RDS_UPGRADE_INTERNAL, nickname = "upgradeRdsByNameInternal")
+    RdsUpgradeV4Response upgradeRdsByClusterNameInternal(@PathParam("workspaceId") Long workspaceId,
+            @NotEmpty @ResourceName @PathParam("name") String clusterName,
+            @QueryParam("targetVersion") TargetMajorVersion targetMajorVersion,
+            @NotEmpty @ValidCrn(resource = CrnResourceDescriptor.USER) @QueryParam("initiatorUserCrn") String initiatorUserCrn);
```
Comment (bergerdenes) [inline core-api/src/main/java/com/sequenceiq/cloudbreak/api/endpoint/v4/stacks/StackV4Endpoint.java:445]: @horadla23 do you think MACHINE_USER is needed here, too?

File: core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/StackV4Controller.java [hunk 5]
```diff
@@ -358,6 +364,13 @@ public FlowIdentifier prepareClusterUpgradeByCrnInternal(Long workspaceId, Strin
         return stackUpgradeOperations.prepareClusterUpgrade(NameOrCrn.ofCrn(crn), restRequestThreadLocalService.getRequestedWorkspaceId(), imageId);
     }
 
+    @Override
+    @InternalOnly
+    public RdsUpgradeV4Response upgradeRdsByClusterNameInternal(Long workspaceId, @NotEmpty @ResourceName String clusterName,
+            TargetMajorVersion targetMajorVersion, @InitiatorUserCrn @NotEmpty @ValidCrn(resource = CrnResourceDescriptor.USER) String initiatorUserCrn) {
```
Comment (bergerdenes) [inline core/src/main/java/com/sequenceiq/cloudbreak/controller/v4/StackV4Controller.java:370]: @horadla23 same here

**Review by bergerdenes — APPROVED** [review]

**Pull request merged**